vpr/src/route/router_lookahead_extended_map.cpp

```diff
@@ -0,0 +1,566 @@
+                                         int to_node_ind,
+                                         float criticality_fac) const {
+    if (from_node_ind == to_node_ind) {
+        return 0.f;
+    }
+
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& rr_graph = device_ctx.rr_nodes;
+
+    RRNodeId to_node(to_node_ind);
+    RRNodeId from_node(from_node_ind);
+
+    int from_x = rr_graph.node_xlow(from_node);
+    int from_y = rr_graph.node_ylow(from_node);
+
+    int to_x = rr_graph.node_xlow(to_node);
+    int to_y = rr_graph.node_ylow(to_node);
+
+    int dx, dy;
+    dx = to_x - from_x;
+    dy = to_y - from_y;
+
+    float extra_delay, extra_congestion;
+    std::tie(extra_delay, extra_congestion) = this->get_src_opin_delays(from_node, dx, dy, criticality_fac);
+
+    int from_seg_index = cost_map_.node_to_segment(from_node_ind);
+    util::Cost_Entry cost_entry = cost_map_.find_cost(from_seg_index, dx, dy);
+
+    if (!cost_entry.valid()) {
+        // there is no route
+        VTR_LOGV_DEBUG(f_router_debug,
+                       "Not connected %d (%s, %d) -> %d (%s)\n",
+                       from_node_ind, device_ctx.rr_nodes[from_node_ind].type_string(), from_seg_index,
+                       to_node_ind, device_ctx.rr_nodes[to_node_ind].type_string());
+        return std::numeric_limits<float>::infinity();
+    }
+
+    float expected_delay = cost_entry.delay + extra_delay;
+    float expected_congestion = cost_entry.congestion + extra_congestion;
+
+    float site_pin_delay = this->get_chan_ipin_delays(to_node);
+    expected_delay += site_pin_delay;
+
+    float expected_cost = criticality_fac * expected_delay + (1.0 - criticality_fac) * expected_congestion;
+
+    VTR_LOGV_DEBUG(f_router_debug, "Requested lookahead from node %d to %d\n", from_node_ind, to_node_ind);
+    const std::string& segment_name = device_ctx.segment_inf[from_seg_index].name;
+    VTR_LOGV_DEBUG(f_router_debug, "Lookahead returned %s (%d) with distance (%zd, %zd)\n",
+                   segment_name.c_str(), from_seg_index,
+                   dx, dy);
+    VTR_LOGV_DEBUG(f_router_debug, "Lookahead delay: %g\n", expected_delay);
+    VTR_LOGV_DEBUG(f_router_debug, "Lookahead congestion: %g\n", expected_congestion);
+    VTR_LOGV_DEBUG(f_router_debug, "Criticality: %g\n", criticality_fac);
+    VTR_LOGV_DEBUG(f_router_debug, "Lookahead cost: %g\n", expected_cost);
+    VTR_LOGV_DEBUG(f_router_debug, "Site pin delay: %g\n", site_pin_delay);
+
+    if (!std::isfinite(expected_cost)) {
+        VTR_LOG_ERROR("infinite cost for segment %d at (%d, %d)\n", from_seg_index, (int)dx, (int)dy);
+        VTR_ASSERT(0);
+    }
+
+    if (expected_cost < 0.f) {
+        VTR_LOG_ERROR("negative cost for segment %d at (%d, %d)\n", from_seg_index, (int)dx, (int)dy);
+        VTR_ASSERT(0);
+    }
+
+    return expected_cost;
+}
+
+// add a best cost routing path from start_node_ind to node_ind to routing costs
```

Comment:
Correct. I added this at @litghost's suggestion to get more samples out of an expansion, since the path gives minimum cost routes to the destination from any of the other nodes in the path.